DuckDuckGo/DuckPlayer/DuckPlayerURLExtension.swift

```diff
@@ -0,0 +1,128 @@
+    private var isYoutubeNoCookie: Bool {
+        host == "www.youtube-nocookie.com" && pathComponents.count == 3 && pathComponents[safe: 1] == "embed"
+    }
+    
+    /// Returns true only if the URL represents a playlist itself, i.e. doesn't have `index` query parameter
+    var isYoutubePlaylist: Bool {
+        guard isYoutubeWatch, let components = URLComponents(url: self, resolvingAgainstBaseURL: false) else {
+            return false
+        }
+
+        let isPlaylistURL = components.queryItems?.contains(where: { $0.name == "list" }) == true &&
+        components.queryItems?.contains(where: { $0.name == "v" }) == true &&
+        components.queryItems?.contains(where: { $0.name == "index" }) == false
+
+        return isPlaylistURL
+    }
+    
+    /// Returns true if the URL represents a YouTube video, but not the playlist (playlists are not supported by Private Player)
+    var isYoutubeVideo: Bool {
+        isYoutubeWatch && !isYoutubePlaylist
+    }
+    
+    /// Attempts extracting video ID and timestamp from the URL. Works with all types of YouTube URLs.
+    var youtubeVideoParams: (videoID: String, timestamp: String?)? {
+        if isDuckURLScheme {
+            guard let components = URLComponents(string: absoluteString) else {
+                return nil
+            }
+            let unsafeVideoID = components.path
+            let timestamp = components.queryItems?.first(where: { $0.name == "t" })?.value
+            return (unsafeVideoID.removingCharacters(in: .youtubeVideoIDNotAllowed), timestamp)
+        }
+
+        if isDuckPlayer {
+            let unsafeVideoID = lastPathComponent
+            let timestamp = getParameter(named: "t")
+            return (unsafeVideoID.removingCharacters(in: .youtubeVideoIDNotAllowed), timestamp)
+        }
+
+        guard isYoutubeVideo,
+              let components = URLComponents(url: self, resolvingAgainstBaseURL: false),
+              let unsafeVideoID = components.queryItems?.first(where: { $0.name == "v" })?.value
+        else {
+            return nil
+        }
+
+        let timestamp = components.queryItems?.first(where: { $0.name == "t" })?.value
+        return (unsafeVideoID.removingCharacters(in: .youtubeVideoIDNotAllowed), timestamp)
+    }
+    
+    
+    var isDuckPlayer: Bool {
+        let isPrivatePlayer = isDuckURLScheme && host == Self.duckPlayerHost
+        return isPrivatePlayer || isYoutubeNoCookie
+        
+    }
+    
+    private func addingTimestamp(_ timestamp: String?) -> URL {
+        guard let timestamp = timestamp,
+              let regex = try? NSRegularExpression(pattern: "^(\\d+[smh]?)+$"),
+              timestamp.matches(regex)
+        else {
+            return self
+        }
+        return appendingParameter(name: "t", value: timestamp)
+    }
+}
+
+extension CharacterSet {
+    static let youtubeVideoIDNotAllowed = CharacterSet(charactersIn: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_").inverted
```

Comment:
Where this data set comes from?

Reply:
That's a very good question.  I have no idea. 

This extension came out directly from macOS implementation and we'll eventually move it to BSK.